src/mono/mono/mini/method-to-ir.c

```diff
@@ -9921,11 +9921,20 @@ mono_method_to_ir (MonoCompile *cfg, MonoMethod *method, MonoBasicBlock *start_b
 			else {
 				klass = NULL;
 				field = mono_field_from_token_checked (image, token, &klass, generic_context, cfg->error);
-				if (!field) {
-					IF_TYPELOAD_ERROR (klass) {
+				if (!field || CLASS_HAS_FAILURE (klass)) {
 						HANDLE_TYPELOAD_ERROR (cfg, klass);
-						break; // reached only in AOT
-					}
+
+						// Reached only in AOT. Cannot turn a token into a class. We silence the compilation error
+						// and generate a runtime exception.
+						if (cfg->error->error_code == MONO_ERROR_BAD_IMAGE)
+							clear_cfg_error (cfg);
+						
+						// We need to push something to the stack for these opcodes.
+						if (il_op == MONO_CEE_LDFLD || il_op == MONO_CEE_LDSFLD || il_op == MONO_CEE_LDFLDA || il_op == MONO_CEE_LDSFLDA) {
+							EMIT_NEW_PCONST (cfg, *sp, NULL);
```

Comment:
This could cause errors since the type might not match.

Reply:
What I mean is that the JIT compilation might fail in ldfld/ldsfld case, since the types on the IL stack don't match whats expected by IL code later.

Reply:
OK, I see.

Reply:
A solution would be to allocate a local of the proper type and push it on the stack.

Reply:
I've been looking into this in `method-to-ir.c`, `MONO_CEE_LDFLD` et al. handler. The call `field = mono_field_from_token_checked (image, token, &klass, generic_context, cfg->error);` is needed to infer the correct type to be pushed onto the stack, but it always returns `NULL` if the class has a failure. There calls inside `mono_field_from_token_checked` that also fail in case of type load failure. So this could be rather involved. 

You mentioned JIT being susceptible here - the macro `HANDLE_TYPELOAD_ERROR` that is invoked if there is a type load error internally calls `TYPE_LOAD_ERROR` in JIT, which immediately fails (== previous behavior). The code after it is only reached in AOT. Does that solve the issue? (Perhaps I should also split up the macro to be more explicit for the JIT case).

Reply:
The same problems could happen in the AOT compiler as well, if the compilation process is continued, it will have a STACK_PTR on the stack which is not what the rest of the IL code expects. Not sure what is the solution here if the field type cannot be determined. The best would be to skip compiling the rest of the IL, but that is very hard to do.

Reply:
The `LDFLD`, `LDSFLD` case now attempts to find out the correct type to push onto the stack, it works on my pet example, but it's not pretty.

Reply:
`MONO_CEE_LDFLD` and `MONO_CEE_LDSFLD` cases now replace the method body with the throw.